draft ttls in marblerun (#49)
edgelesssys/marblerun.sh: content/docs/workflows/define-manifest.md

                    "MARBLE_CERT": "{{ pem .Marblerun.MarbleCert.Cert }}",
                    "MARBLE_KEY": "{{ pem .Marblerun.MarbleCert.Private }}"
                }
            },
            "TLS": [

>                 "front", "other"

Reviewer: Why don't we name them similar to the service?

Author: If I understand correctly service meaning marble?
I wanted to emphasize that the names of the tags are arbitrary and don't need to exactly match the services.

Reviewer 2: Maybe change the labels from `front` and `back` to `frontend_tls` and `backend_first_tls`?
This way we keep the naming from the service, but also show that the names don't need to match

Reviewer 3: TLS tags/labels should be named after what they provide, not after the marble(s) that use them.

Author: I'm kind of agree and disagree. IMO, the most probable scenario is something like 5 gRPC Services and maybe 1 web service. Now we maybe want to share the configuration of the gRPC services.  Even if they all listen on the same port, and the Incoming definition can be shared, we would still need to define each service in the Outgoing section. Therefore, we also define that Service1 connection to Service1 should be wrapped, which makes little sense. I see that it is nice to have all the services that talk to each other in one list (e.g. the Outgoing part), but this overdefinition can quickly explode, e.g. when Service1 only talks to Service2 and Service2 only to Service3. Seeing those services grouped as one TLS definition makes me think that Service1 talks to every service in the list.  Again, assuming they all use the same incoming port. Therefore, my way of configuring would be marble-wise, if there is no good reason not to. Maybe I would even define an "gRPC_incoming" tag, use that for each marble and define the Outgoing part separately for each.  

Furthermore, it is quite hard to show the different variations with 2 of them. 

tl;dr: 2 Options: By default TLS labels define 
1. marbles (1-to-1)  
2. "groups" of marbles (1-to-n)

Reviewer 3: I don't get all of what you say, but doesn't matter.

> we would still need to define each service in the Outgoing section

You can have an Outgoing rule with a wildcard address and fixed port. One could argue that a wildcard address is always preferable regarding security. At least it's not dangerous.

> Maybe I would even define an "gRPC_incoming" tag

Yes, that's my point.

In general, the feature is designed to support n-to-m. (One reason is avoiding duplicate entries in multiple TLS definititions.)

Anyway, I didn't necessarily want anything to be changed in this example, but just clarify to all participants that "name them similar to the service" is not the rule of thumb.

awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' public_key.pem
```

## Manifest:TLS

> The TLS entry holds a list of tags which can be used in a marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. Outgoing connections are defined by `Port` and `Addr`. Incoming connections just use `Port` and always define `Addr` as `"*"`. Both domains and IPs can be used as `Addr`. Previously defined certificates can be referenced by name. This disables client authentication.

Author: Should we warn about the disabling of client authentication in a highlighted box?

> The TLS entry holds a list of tags which can be used in a Marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. Outgoing connections are defined by `Port` and `Addr`. Incoming connections just use `Port` and always define `Addr` as `"*"`. Both domains and IPs can be used as `Addr`. Previously defined certificates can be referenced by name. This disables client authentication.

Reviewer: Why do previously defined certificates disable client authentication?
Wouldn't it be better to have an explicit parameter for that?
`domains` are a bit ambiguous and probably require one more sentence on DNS names.

Author: Currently, client authentication is automatically disabled as specified in https://github.com/edgelesssys/emojivoto/pull/11#issuecomment-790501876. This would only need a small change in marblerun to be able to configure it manually. There are probably use cases where client authentication can/must be enabled. If there are no objections, I will add this feature. 

I'll also add some more information on the `domains` part, here. Especially that the internal kubernetes domains can be used.

awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' public_key.pem
```

## Manifest:TLS

> The TLS entry holds a list of tags which can be used in a Marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. Outgoing connections are defined by `Port` and `Addr`. Incoming connections just use `Port` and always define `Addr` as `"*"`. Both domains and IPs can be used as `Addr`. Most importantly, the DNS names of other services can be used.  Previously defined certificates can be referenced by name. This disables client authentication.

Reviewer 3: > Incoming connections just use `Port` and always define `Addr` as `"*"`.

So why do they define `Addr` at all?

Author: We could leave it out and define it in marblerun ourselves. We also could leave it there to remind the user that we don't bind on a specific interface but all interfaces.

Reviewer 3: Sorry, I thought we already support leaving it out. So is the following valid or not?
https://github.com/edgelesssys/marblerun/blob/3891ebcbf717e51218c9c3bb3b66270f6dd9713b/test/manifests.go#L152-L155

Author: Sorry, I had to look into the marblerun implementation again. Currently, your example is already supported. I have no idea why I defined Addr as "*". So I will just remove the sentence and replace it with something like: "Incoming connections just use `Port` and which will be used on all interfaces."

> The TLS entry holds a list of tags which can be used in a Marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. Outgoing connections are defined by `Port` and `Addr`. Incoming connections just use `Port` and always define `Addr` as `"*"`. Both domains and IPs can be used as `Addr`. Most importantly, the DNS names of other services can be used.  Previously defined certificates can be referenced by name. This disables client authentication.

Reviewer 3: This paragraph should be split in two or three
* Short intro and common facts (if any)
* Incoming connections: Why, when, and how to use them (internal mutual vs external)
* Outgoing connections: Why, when, and how to use them

```javascript
{
    //...
    "TLS":
    {
        "frontend_tls_1": {
            "Outgoing": [
                {
                    "Port": "8080",
                    "Addr": "service.name"
                },
                {
                    "Port": "4443",
                    "Addr": "127.0.0.1"

>                 }

Reviewer 3: I think this is a bad example in the CC context.

            ],
            "Incoming": [
                {
                    "Port": "8080",
                    "Addr": "*",
                    "Cert": "rsa_cert",
                    "ClientAuth": false

>                 }

Reviewer 3: We should also show how to define an incoming connection for the most common case, i.e., internal mutual TLS.

awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' public_key.pem
```

## Manifest:TLS

The TLS entry holds a list of tags which can be used in a Marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. To elevate the connection between two marbles to TLS, the client needs to set the server under `Incoming` and the server needs to define its service under `Incoming`.

> Outgoing connections are defined by `Port` and `Addr`. Both domains and IPs can be used as `Addr`. Most importantly, the DNS names of other services can be used.

Reviewer 3: ```suggestion
Outgoing connections are defined by `Port` and `Addr`. For `Addr`, you can use both IP addresses and domains, e.g., the DNS names of other services.
```
prefer active voice where possible (if it doesn't complicate the sentence or sound odd)

> Incoming connections are listened to on all interfaces and are only defined by `Port`. For services that are used by external clients, you must set `Cert`. Previously defined certificates can be referenced here by name. Client authentication then needs to be disabled by setting `DisableClientAuth` to `true`.

Reviewer 3: ```suggestion
Incoming connections are defined by `Port`. All interfaces will be intercepted.. For services that are used by external clients, you must set `Cert`. Previously defined certificates can be referenced here by name. Client authentication then needs to be disabled by setting `DisableClientAuth` to `true`.
```
"Incoming connections are listened to on all interfaces" reads odd. Maybe even wrong as it may suggest that TTLS controls on what interfaces to listen. However, my suggestion still is hard to understand without more context. Does it even add value to mention this at all? May be more confusing than helpful to most readers.

> Incoming connections are listened to on all interfaces and are only defined by `Port`. For services that are used by external clients, you must set `Cert`. Previously defined certificates can be referenced here by name. Client authentication then needs to be disabled by setting `DisableClientAuth` to `true`.

Reviewer 3: ```suggestion
Incoming connections are listened to on all interfaces and are only defined by `Port`. For services used by external clients, you must disable client authentication by setting `DisableClientAuth` to `true` and set `Cert`. Use the name of a certificate defined in the [Secrets section](#manifestsecrets).
```
I tried to make the reason for disabling client auth more clear by (grammatically) binding it to "external clients".
